xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/internal/doc/XWikiAttachmentList.java

```diff
@@ -0,0 +1,240 @@
+     * @since 10.0RC1
+     */
+    @Override
+    public boolean add(XWikiAttachment attachment)
+    {
+        return set(attachment) != null;
+    }
+
+    /**
+     * Adds attachment to the list in order of filename.
+     * 
+     * @param index index is ignored as list is reordered based on filename
+     * @param attachment XWikiAttachment to add to the list
+     * @since 10.0RC1
+     */
+    @Override
+    public void add(int index, XWikiAttachment attachment)
+    {
+        add(attachment);
+    }
+
+    @Override
+    public void clear()
+    {
+        map.clear();
+        updatedMap();
+    }
+
+    /**
+     * Adds all attachments to the list in order of filename.
+     * 
+     * @param index index is ignored as list is reordered based on filename
+     * @param c Collection that contains XWikiAttachment objects
+     * @since 10.0RC1
+     */
+    @Override
+    public boolean addAll(Collection<? extends XWikiAttachment> c)
+    {
+        boolean changed = false;
+        for (XWikiAttachment x : c) {
+            if (map.put(x.getFilename(), x) == null) {
+                changed = true;
+                added(x);
+            }
+        }
+        if (changed) {
+            updatedMap();
+        }
+        return changed;
+    }
+
+    /**
+     * Adds all attachments to the list in order of filename.
+     * 
+     * @param index index is ignored as list is reordered based on filename
+     * @param c Collection that contains XWikiAttachment objects
+     * @since 10.0RC1
+     */
+    @Override
+    public boolean addAll(int index, Collection<? extends XWikiAttachment> c)
+    {
+        return addAll(c);
+    }
+
+    @Override
+    public XWikiAttachment remove(int index)
+    {
+        XWikiAttachment removedAttachment = map.remove(this.decorated().get(index).getFilename());
+        if (removedAttachment != null) {
+            updatedMap();
+        }
+        return removedAttachment;
+    }
+
+    /**
+     * Removes XWikiAttachment.
+     * 
+     * @param attachment XWikiAttachment to remove.
+     * @return true unless the attachment is not found
+     * @since 10.0RC1
+     */
+    @Override
+    public boolean remove(Object attachment)
+    {
+        XWikiAttachment xwikiAttachment = (XWikiAttachment) attachment;
+        if (map.remove(xwikiAttachment.getFilename()) == null) {
+            return false;
+        }
+        updatedMap();
+        return true;
+    }
+
+    /**
+     * Adds or replaces attachment with the same filename as the parameter.
+     * 
+     * @param attachment the attachment to add to the list
+     * @return the attachment that was previously matched to the same filename or null if no attachment was matched to
+     *         it
+     * @since 10.0RC1
+     */
+    public XWikiAttachment set(XWikiAttachment attachment)
+    {
+        XWikiAttachment put = map.put(attachment.getFilename(), attachment);
+        if (put == null || !put.equals(attachment)) {
```

Comment:
I would use == instead of equals. XWikiAttachment don't implement equals and even if they were is might be super slow (they would have to compare sometime big files contents).